src/drive/web/modules/drive/Toolbar/components/DocumentQualification.jsx

```diff
@@ -133,42 +143,109 @@ DocumentCategory.propTypes = {
     }
+    this.textInput = React.createRef()
   }
 
+  getFilenameFromCategory = (item, t) => {
+    const realItem = getItemById(item.itemId)
+    if (realItem) {
+      const name = t(`Scan.items.${realItem.label}`)
+      return `${name.replace(/ /g, '-')}_${new Date()
+        .toLocaleDateString()
+        .replace(/:/g, '-')
+        .replace(/\//g, '-')}`
+    } else {
+      return this.defaultFilename
+    }
+  }
   onSelect = item => {
-    this.setState({ selected: item })
+    const { t } = this.props
+    const { shouldAutomaticalyRenameFile } = this.state
+    let filename = null
+    if (shouldAutomaticalyRenameFile) {
+      filename = this.getFilenameFromCategory(item, t)
+    } else {
+      filename = this.state.filename
+    }
+    this.setState({ selected: item, filename })
+
     const { onQualified } = this.props
     if (onQualified) {
       const realItem = getItemById(item.itemId)
-      //We only call the callback if a "real" item is selected
-      //not if `Scan.categories.undefined` is
-      if (realItem) onQualified(realItem)
+      onQualified(
+        realItem ? realItem : undefined,
+        filename + filename_extension
+      )
     }
   }
 
   render() {
-    const { t } = this.props
-    const { selected } = this.state
+    const { t, title, editFileName } = this.props
+    const { selected, filename, shouldAutomaticalyRenameFile } = this.state
     return (
       <MuiCozyTheme>
-        <Title className="u-mv-1">{t('Scan.doc_type')}</Title>
+        {editFileName && (
+          <>
+            <Label htmlFor="filename_input">{t('Scan.filename')}</Label>
+            <InputGroup
+              fullwidth
+              append={<span className="u-pr-1">{filename_extension}</span>}
+              className="u-bdrs-3"
+            >
+              <Input
+                placeholder={t('Scan.filename')}
+                value={filename}
+                onChange={event => {
+                  //If the user write something once, we don't want to rename the file automatically anymore
+                  if (shouldAutomaticalyRenameFile) {
+                    this.setState({ shouldAutomaticalyRenameFile: false })
+                  }
+                  if (event.target.value === '')
+                    this.setState({ shouldAutomaticalyRenameFile: true })
+                  this.setState({
+                    filename: event.target.value
+                  })
+                }}
+                onClick={() => {
+                  if (shouldAutomaticalyRenameFile)
+                    this.textInput.current.setSelectionRange(0, filename.length)
+                }}
+                onBlur={() => {
+                  if (filename === '')
+                    this.setState({
+                      filename: this.getFilenameFromCategory(selected, t)
+                    })
+                }}
+                inputRef={this.textInput}
+                id="filename_input"
```

Comment:
A constant would be good for that, so that we don't accidentally change the input but not the label.